LMMS/lmms
# MIDI Program Change message support in VeSTige, SF2 Player and GIG Player

Adds the native support for the following MIDI events in VeSTige:
- Program Change
- Controller Change for controllers 0 and 32, which are Bank Select MSB and LSB, respectively

The patch adds two toggle switches to VeSTIge GUI:
- Capture Program Change events
- Use Bank Select LSB
which are loadable and saveable. To make the room for new switches it was neccessary to reorganize slightly the existing user interface - "Show/Hide GUI" and "Stop all notes" buttons were reduced to the form of icons. Additionally, the MIDI preset number was moved from preset name line to the "Preset:" caption above it, so that there is more space for the actual preset name.

**Description of new features**

When "Capture Program Change events" switch is on, the events above are captured directly by VeSTige and used to determine the ID of preset to switch to; the VST plugin will not receive these events. When the switch is off the MIDI events will be passed through to the plugin.

When "Use Bank Select LSB" switch is on, both controllers 0 and 32 are used. The values are combined to obtain the bank number between 0 and 16383 (2^14 - 1). When the switch is off, only controller 0 is used and is treated as LSB. This allows to choose banks between 0 and 127 (2^7 - 1). This allows to work with MIDI devices that use either of the approaches.

The target preset ID is determined from the bank number and the program number. Presets 0 to 127 belong to bank 0, presets 128..255 fall into bank 1, presets 256..383 belong to bank 2, and so on.

Reception of the Program Change and/or Bank Select messages will trigger the update of the VeSTige GUI, so the preset name is visible immediately.

**Validation of changes**
_Preparation_
- Connect any MIDI controller capable of sending subject messages (e. g. Alesis QX49)
- Add VeSTige to the song editor and load any VST plugin that has its GUI and at least a few presets. I have used SuperWave P8 and SQ8L.
- Enable the MIDI input. If you use ALSA Sequencer as MIDI interface, pick appropriate MIDI device from the list.
- Open the VST GUI window

_Validating Program Change_
- Having both switches disabled, trigger Program Change event at your controller.
  - if the plugin you loaded accepts Program Change messages (e. g. SuperWave P8), you will see a change in the plugin's GUI, but no change of the preset in VeSTige
  - if the plugin does not handle Program Change (e. g. SQ8L), there will be no change in the GUI
- Enable "Capture Program Change events" switch and trigger Program Change event.
  - the preset name displayed in VeSTige GUI should change according to the program number received, contents of plugin's GUI should be updated as well

_Validating Bank Select_
Note: here I recommend using a VST plugin with more than 128 presets.
- Enable "Use Bank Select LSB" switch
- Send Program Change with the value of 0
  - the first preset should be selected
- Send Bank Select MSB with the value of 0 and Bank Select LSB with the value of 1
  - the 129th preset should be selected (bank 1, program 0, meaning Preset ID=128)
- Send Bank Select LSB with the value of 2
  - the 257th preset should be selected (bank 2, program 0, meaning Preset ID=256)
- Send Bank Select MSB with the value of 1 and Bank Select LSB with the value of 0
  - the last preset should be selected (bank 128, program 0, meaning Preset ID=16384)

- Disable "Use Bank Select LSB" switch
- Send Bank Select MSB with the value of 0 and Bank Select LSB with the value of 1
  - the first preset should be selected (bank 0, program 0, meaning Preset ID=0)
- Send Bank Select LSB with any value
  - there should be no reaction of the plugin
- Send Bank Select MSB with the value of 1
  - the 129th preset should be selected (bank 1, program 0, meaning Preset ID=128)

_Validating save & load_
- Set any state of the switches in VeSTige GUI, then save the project
- Open a new project
- Reopen the previously saved project
- Open the GUI of the instance of VeSTige loaded into the project
  - the state of switches should be the same as when the project was saved

**Notes**
Change in scope of #5138 (already closed) which is actually a part of #1472.
The pull request includes a fix for #5139 which has not yet been merged to the master branch, but is required by the new feature.
